In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import scipy.stats as stats
import math
pd.options.display.max_columns = None

In [2]:
df_sub = pd.read_pickle('/Users/reesezhuang/Desktop/NYU Stern/Final individual/subscribers')

In [3]:
df_sub

subid package_type  num_weekly_services_utilized preferred_genre  \
162502  21724479      economy                           NaN          comedy   
380374  23383224         base                           NaN          comedy   
496617  26844789     enhanced                      3.362885        regional   
356799  29417030         base                           NaN           drama   
681701  26723159         base                      3.731089          comedy   
...          ...          ...                           ...             ...   
509326  21434712     enhanced                      2.594078          comedy   
229587  25843074     enhanced                      2.361370          comedy   
281340  24799085         base                           NaN          comedy   
87198   21308040         base                           NaN          comedy   
598241  20166335         base                           NaN          comedy   

                       intended_use  weekly_consumption_hour  \
162502  access to exclusive content                      NaN   
380374  access to exclusive content                22.951579   
496617                  replace OTT                36.001186   
356799                  replace OTT                20.051667   
681701                  replace OTT                22.951579   
...                             ...                      ...   
509326               supplement OTT                28.751405   
229587                  replace OTT                27.301448   
281340  access to exclusive content                31.651317   
87198   access to exclusive content                      NaN   
598241                  replace OTT                25.851492   

        num_ideal_streaming_services  retarget_TF   age male_TF country  \
162502                           NaN        False   NaN   False     UAE   
380374                           NaN        False  70.0   False     UAE   
496617                      1.951776        False  25.0    True     UAE   
356799                           NaN        False  30.0   False     UAE   
681701                      2.506384        False  28.0   False     UAE   
...                              ...          ...   ...     ...     ...   
509326                      1.905485        False  38.0   False     UAE   
229587                      1.835189         True  49.0   False     UAE   
281340                           NaN        False  45.0   False     UAE   
87198                            NaN        False   NaN   False     UAE   
598241                           NaN        False  55.0   False     UAE   

       attribution_technical attribution_survey   op_sys  \
162502              facebook           facebook  Android   
380374              facebook           facebook      NaN   
496617               organic           facebook      iOS   
356799                search                 tv  Android   
681701             discovery            youtube      iOS   
...                      ...                ...      ...   
509326              facebook   facebook_organic      iOS   
229587        google_organic           referral      iOS   
281340              facebook           facebook      iOS   
87198               facebook           facebook      iOS   
598241               organic                 tv      iOS   

        months_per_bill_period              plan_type  monthly_price  \
162502                       4  base_uae_14_day_trial         4.7343   
380374                       4  base_uae_14_day_trial         4.7343   
496617                       4  base_uae_14_day_trial         4.7343   
356799                       4  base_uae_14_day_trial         4.7343   
681701                       4  base_uae_14_day_trial         4.7343   
...                        ...                    ...            ...   
509326                       4  base_uae_14_day_trial         4.7343   
229587                       4  base_uae_14_day_trial         4.7343   
281340                 

In [4]:
df_sub.describe()

subid  num_weekly_services_utilized  weekly_consumption_hour  \
count  2.276280e+05                 117178.000000            189698.000000   
mean   2.498624e+07                      2.998568                27.997724   
std    2.885543e+06                      0.748752                 4.976341   
min    2.000001e+07                      0.049049               -32.146760   
25%    2.248998e+07                      2.493926                24.401536   
50%    2.497011e+07                      2.760506                27.301448   
75%    2.749020e+07                      3.271574                30.201361   
max    2.999998e+07                     14.335378                76.599962   

       num_ideal_streaming_services           age  months_per_bill_period  \
count                 115458.000000  1.924590e+05                227628.0   
mean                       1.999848  7.579755e+02                     4.0   
std                        0.249508  1.920204e+05                     0.0   
min                       -0.919933  0.000000e+00                     4.0   
25%                        1.867354  3.500000e+01                     4.0   
50%                        1.908159  4.600000e+01                     4.0   
75%                        1.953559  5.700000e+01                     4.0   
max                        6.955714  8.172000e+07                     4.0   

       monthly_price  discount_price  creation_until_cancel_days  \
count  227628.000000   227628.000000               151768.000000   
mean        4.732049        4.512188                   16.570832   
std         0.101049        0.097078                   35.777986   
min         0.807400        0.770700                 -108.000000   
25%         4.734300        4.514100                    2.000000   
50%         4.734300        4.514100                    8.000000   
75%         4.734300        4.514100                   13.000000   
max         5.101300        5.027900                  312.000000   

         revenue_net       join_fee  
count  192724.000000  192724.000000  
mean        1.557353       0.115145  
std         3.068829       0.176971  
min        -5.982100      -0.660600  
25%         0.000000       0.036700  
50%         0.000000       0.036700  
75%         2.532300       0.110100  
max        15.634200       0.734000

In [5]:
df_sub.isnull().sum()

subid                                0
package_type                     35574
num_weekly_services_utilized    110450
preferred_genre                  36326
intended_use                      3549
weekly_consumption_hour          37930
num_ideal_streaming_services    112170
retarget_TF                          0
age                              35169
male_TF                            269
country                              0
attribution_technical                0
attribution_survey                2644
op_sys                           13375
months_per_bill_period               0
plan_type                            0
monthly_price                        0
discount_price                       0
account_creation_date                0
creation_until_cancel_days       75860
cancel_before_trial_end              0
trial_end_date                       0
initial_credit_card_declined         0
revenue_net                      34904
join_fee                         34904
language                 

In [6]:
df_sub['month'] = df_sub['account_creation_date'].map(lambda x: str(x)[:7])

In [7]:
df_sub['month'].value_counts()

2020-02    28089
2019-11    26828
2020-03    26122
2019-07    25708
2019-08    25434
2019-12    25339
2020-01    24843
2019-10    22412
2019-09    20190
2019-06     2663
Name: month, dtype: int64

In [8]:
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "base_uae_14_day_trial", "base_14_day_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "high_uae_14_day_trial", "high_14_day_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "low_uae_no_trial", "low_no_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "base_uae_14_day_trial", "base_14_day_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "high_sar_14_day_trial", "high_14_day_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "low_gbp_14_day_trial", "low_14_day_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "high_aud_14_day_trial", "high_14_day_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "low_eur_no_trial", "low_no_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "low_sar_no_trial", "low_no_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "base_eur_no_trial_7_day_guarantee", "base_no_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "high_jpy_14_day_trial", "high_14_day_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "base_eur_14_day_trial", "base_14_day_trial", df_sub["plan_type"])
df_sub["plan_type"] = np.where(df_sub["plan_type"] == "base_uae_no_trial_7_day_guarantee", "base_no_trial", df_sub["plan_type"])
df_sub['plan_type'].value_counts()

base_14_day_trial    227114
high_14_day_trial       340
low_no_trial            169
low_14_day_trial          4
base_no_trial             1
Name: plan_type, dtype: int64

In [9]:
df_sub = df_sub[df_sub['plan_type'] != 'base_no_trial']
df_sub = df_sub[df_sub['plan_type'] != 'low_14_day_trial']
df_sub['plan_type'].value_counts()

base_14_day_trial    227114
high_14_day_trial       340
low_no_trial            169
Name: plan_type, dtype: int64

In [10]:
#Drop unreasonable data
#Target customers aged between 7 and 80
df_sub.drop(df_sub[df_sub['weekly_consumption_hour'] < 0 ].index, inplace=True)
df_sub.drop(df_sub[df_sub['num_ideal_streaming_services'] < 0 ].index, inplace=True)
df_sub.drop(df_sub[df_sub['age'] > 80 ].index, inplace=True)
df_sub.drop(df_sub[df_sub['age'] < 7 ].index, inplace=True)
df_sub.drop(df_sub[df_sub['join_fee'] < 0 ].index, inplace=True)

# These different pricing schemes will be compared:
## 1. base monthly price with a 14-days trial (all currencies)
## 2. low monthly price without trial (all currencies)
## 3. high monthly price with a 14-days trial (all currencies)

In [11]:
df_sub['plan_type'].value_counts()

base_14_day_trial    225724
high_14_day_trial       339
low_no_trial            168
Name: plan_type, dtype: int64

In [12]:
# conversion defined as customer who made past trial, didn't cancel after trial, and had revenue > 1
df_sub['conversion'] = np.where((df_sub['cancel_before_trial_end'] == False) | (df_sub['revenue_net'] < 1) |
                                (df_sub['refund_after_trial_TF'] == True),
                                    False, True)

In [13]:
df_sub[df_sub['plan_type'] == 'low_no_trial']['month'].value_counts() 

2019-07    146
2019-11      8
2019-08      6
2019-10      4
2019-09      4
Name: month, dtype: int64

In [14]:
df_sub[df_sub['plan_type'] == 'base_14_day_trial']['month'].value_counts()

2020-02    27930
2019-11    26326
2020-03    26041
2019-07    25365
2019-08    25281
2019-12    25143
2020-01    24665
2019-10    22293
2019-09    20041
2019-06     2639
Name: month, dtype: int64

In [15]:
df_sub[df_sub['plan_type'] == 'high_14_day_trial']['month'].value_counts()

2019-11    324
2019-07      5
2019-08      3
2019-10      2
2019-09      2
2020-01      2
2020-03      1
Name: month, dtype: int64

### Pick a same time period with no chance of data still in trial period corrupting calculations.
### So, 2019-07 through 2020-02

# First let's compare base and low

### Null hypothesis: convert_low = convert_base
### Alternative hypothesis: convert_low > convert_base

In [16]:
df_ab1 = df_sub[df_sub['plan_type'] != 'high_14_day_trial'] #So select only base and low
date_list = ["2019-07", "2019-08", "2019-09", "2019-10", "2019-11", "2019-12", "2020-01", "2020-02"]
df_ab1 = df_ab1[df_ab1['month'].isin(date_list)]
df_ab1["month"].value_counts()

2020-02    27930
2019-11    26334
2019-07    25511
2019-08    25287
2019-12    25143
2020-01    24665
2019-10    22297
2019-09    20045
Name: month, dtype: int64

In [17]:
df_ab1.groupby("plan_type")["conversion"].value_counts()

plan_type          conversion
base_14_day_trial  False         125925
                   True           71119
low_no_trial       True             130
                   False             38
Name: conversion, dtype: int64

In [18]:
df_ab1_count = df_ab1.groupby(['plan_type'])['conversion'].agg(['count','sum']).reset_index()
df_ab1_count['conversion'] = df_ab1_count['sum'] / df_ab1_count['count']
df_ab1_count

plan_type   count    sum  conversion
0  base_14_day_trial  197044  71119     0.36093
1       low_no_trial     168    130     0.77381

In [19]:
df_ab1.groupby('plan_type').discount_price.mean()

plan_type
base_14_day_trial    4.514079
low_no_trial         1.026945
Name: discount_price, dtype: float64

In [20]:
p0 = df_ab1_count.loc[0,'conversion']
p1 = df_ab1_count.loc[1,'conversion']
print("p0: %.2f" % p0)
print("p1: %.2f" % p1)

p0: 0.36
p1: 0.77


In [21]:
# Z critical value with 95% confidence level
from scipy.stats import norm
norm.ppf(1-(0.05/1))

1.6448536269514722

In [22]:
def z_score(p1:float, p0:float, n = int)->float:
    z = (p1 - p0) / math.sqrt(p0*(1-p0)/n)
    return z

In [23]:
z_score(p1,p0,df_ab1_count.loc[1,'count'])

11.142762810119882

## Because 11.14 > 1.64, we reject the null hypothesis and low monthly price without trial scheme was effective and increased conversion rate.

# Secondly, let's compare base and high

### Null hypothesis: convert_high = convert_base
### Alternative hypothesis: convert_high > convert_base

In [24]:
df_ab2 = df_sub[df_sub['plan_type'] != 'low_no_trial'] #So select only base and low
df_ab2 = df_ab2[df_ab2['month'].isin(date_list)]
df_ab2["month"].value_counts()

2020-02    27930
2019-11    26650
2019-07    25370
2019-08    25284
2019-12    25143
2020-01    24667
2019-10    22295
2019-09    20043
Name: month, dtype: int64

In [25]:
df_ab2.groupby("plan_type")["conversion"].value_counts()

plan_type          conversion
base_14_day_trial  False         125925
                   True           71119
high_14_day_trial  False            238
                   True             100
Name: conversion, dtype: int64

In [26]:
df_ab2_count = df_ab2.groupby(['plan_type'])['conversion'].agg(['count','sum']).reset_index()
df_ab2_count['conversion'] = df_ab2_count['sum'] / df_ab2_count['count']
df_ab2_count

plan_type   count    sum  conversion
0  base_14_day_trial  197044  71119    0.360930
1  high_14_day_trial     338    100    0.295858

In [27]:
df_ab2.groupby('plan_type').discount_price.mean()

plan_type
base_14_day_trial    4.514079
high_14_day_trial    4.987725
Name: discount_price, dtype: float64

In [28]:
p0 = df_ab2_count.loc[0,'conversion']
p1 = df_ab2_count.loc[1,'conversion']
print("p0: %.2f" % p0)
print("p1: %.2f" % p1)

p0: 0.36
p1: 0.30


In [29]:
z_score(p1,p0,df_ab2_count.loc[1,'count'])

-2.4909437429665013

## Because -2.49 < 1.64, we fail to reject the null hypothesis and high monthly price with a 14-days trial scheme was not effective and cannot increase conversion rate.